gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -360,7 +360,7 @@ public WritableByteChannel create(final StorageResourceId resourceId, CreateObje
     Optional<Long> overwriteGeneration =
         resourceId.hasGenerationId()
             ? Optional.of(resourceId.getGenerationId())
-            : Optional.of(getWriteGeneration(resourceId, options.overwriteExisting()));
+            : getWriteGeneration(resourceId, options.overwriteExisting());
```

Comment:
Seems like we don't need to check existence of the object here and just depend on `options.overwriteExisting()` condition:
```java
options.overwriteExisting() ? Optional.absent() : Optional.of(0);
```

Reply:
We do need to take care of the file-already-exists case, in which chase we want to set some generation ID if overwrite is false, I think better to just consolidate into one method.

Reply:
This sounds about right but getWriteGeneration() method actually fails request directly without making request to GCS if overwrite set to false, I think better to just consolidate into one method and save one request.

Here I tend to maintain the original logic as much as possible before I am clear about the impact of set/unset ifMatchGeneration on other requests.

Reply:
It actually does not save a request, because `getWriteGeneration()` always makes at least couple requests in the `getItemInfo(resourceId)` method.

I guess the only edge case, is if this logic somehow allows to handle directory overwrite with file, i.e. what happens when you write `gs://a/b/c` file if directory `gs://a/b/c/` already exists? If it's not handled, then we can just remove `getWriteGeneration()` altogether (if it used only here).

Reply:
This method is also used by compose method as well. Removal of it causes some irrelevant tests to fail. 

About the edge case you mentioned we actually handle this case - by failing the attempt - at the beginning stage of the creation, so we have this check`checkNoDirectoryConflict` in create method of GoogleCloudStorageFileSystem.

Reply:
Maybe compose has the same bug?

Regarding `create` changes in this PR, because it sets existing object generation in overwrite request, it means that it can not overwrite existing objects concurrently, so it does not fix all the use cases for overwrite.

Reply:
I'll leave compose as is - always trying to set generation, for now for this PR.

With that said, I will let generateWriteGeneration merely do what it means to do - generate generation, rather than mixed with other responsibilities like check overwrite and throw exception etc. For creation it's now updated more like what you suggested earlier but remain assigning generation# for non-overwrite case.